fiftyone/operators/decorators.py

```diff
@@ -93,3 +96,125 @@ def dir_state(dirpath):
+            If not provided, the function arguments are used as the key by serializing them as JSON.
+        store_name (None): Custom name for the execution store.
+            Defaults to the function name.
+
+    Example Usage:
+        # Standalone function with default caching
+        @execution_cache(ttl=60, store_name="query_cache")
+        def expensive_query(ctx, path):
+            return ctx.dataset.count_values(path)
+
+        # Instance method with dataset-scoped caching
+        class Processor:
+            @execution_cache(ttl=60, store_name="processor_cache")
+            def expensive_query(self, ctx, path):
+                return ctx.dataset.count_values(path)
+
+        # Using a custom key function
+        def custom_key_fn(ctx, path, user_id):
+            return [path, user_id]
+
+        @execution_cache(ttl=90, key_fn=custom_key_fn)
+        def user_specific_query(ctx, path, user_id):
+            return ctx.dataset.count_values(path)
+    """
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            ctx, ctx_index = _get_ctx_from_args(args)
+
+            if ctx_index == -1 or ctx_index > 1:
+                raise ValueError(
+                    f"@execution_cache requires the first argument to be an ExecutionContext"
+                )
+
+            dataset_id = ctx.dataset._doc.id if link_to_dataset else None
+
+            # Store name
+            resolved_store_name = store_name or func.__name__
+            if not isinstance(resolved_store_name, str):
+                raise ValueError("`store_name` must be a string if provided.")
+
+            # Cache key
+            cache_key_list = args[
+                ctx_index + 1 :
+            ]  # Skip `ctx` and `self` if present
+            if key_fn:
+                try:
+                    cache_key_list = key_fn(*args, **kwargs)
+                    if not isinstance(cache_key_list, list):
+                        raise ValueError(
+                            "Custom key function must return a list."
+                        )
+                except Exception as e:
+                    raise ValueError(
+                        f"Failed to create custom cache key `{func.__name__}`: {e}"
+                    )
+
+            cache_key = str(hashkey(_convert_args_to_dict(cache_key_list)))
+
+            store = ExecutionStore.create(
+                store_name=resolved_store_name, dataset_id=dataset_id
+            )
+
+            cached_value = store.get(cache_key)
+            if cached_value is not None:
+                return cached_value  # Cache hit
+
+            result = func(*args, **kwargs)
+            store.set(cache_key, result, ttl=ttl)
+            return result
+
+        return wrapper
+
+    return decorator
+
+
+def _get_ctx_from_args(args):
+    ctx_index = -1
+    for i, arg in enumerate(args):
+        if hasattr(arg, "dataset"):
+            ctx_index = i
+            break
```

Comment:
i'm kind of new to this but is `hasattr(arg, "dataset")` heuristic sufficient?

Reply:
`isinstance(arg, ExecutionContext)` would be better....this is only to allow simpler mocks in the test.

Reply:
Now using `isinstance` much better!